File: AndroidApp/app/src/androidTest/java/com/ep133/sampletool/robots/KitRobot.kt
package com.ep133.sampletool.robots

import androidx.compose.ui.test.assert
import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.hasAnyAncestor
import androidx.compose.ui.test.hasTestTag
import androidx.compose.ui.test.hasText
import androidx.compose.ui.test.junit4.ComposeContentTestRule
import androidx.compose.ui.test.performClick
import androidx.compose.ui.test.performScrollTo
import com.ep133.sampletool.ui.TestTags

/**
 * Robot for the SAMPLES (Kit) screen — the loop chopper's chrome plus a smoke check that KIT
 * mode renders the embedded Kit Builder. Kit Builder's own internals (pack browser, pad canvas)
 * are out of scope here; see a future KitBuilderRobot for that.
 */
class KitRobot(rule: ComposeContentTestRule) : BaseRobot(rule) {

    /** The chop-mode slice counter is the screen's stable unique anchor. */
    fun assertModeChipsVisible() = apply {
        text("SLICE COUNT").assertIsDisplayed()
    }

    // Mode toggle — Ep133SectionLabel uppercases its text, so match the rendered case.
    fun selectChopMode() = apply { tag(TestTags.KIT_MODE_CHOP).performClick() }
    fun selectKitMode() = apply { tag(TestTags.KIT_MODE_KIT).performClick() }
    fun assertChopModeVisible() = apply { text("LOOP CHOPPER").assertIsDisplayed() }
    fun assertKitModeVisible() = apply { text("KIT BUILDER").assertIsDisplayed() }

    // Group + choke bar (shared with Pads via TestTags.groupChip)
    fun selectGroup(group: String) = apply { tag(TestTags.groupChip(group)).performClick() }
    fun toggleChoke() = apply { tag(TestTags.GROUP_CHOKE_TOGGLE).performClick() }
    fun assertChokeOn() = apply { text("ON").assertIsDisplayed() }
    fun assertChokeOff() = apply { text("OFF").assertIsDisplayed() }

    // Slice count selector
    fun incrementSliceCount() = apply { tag(TestTags.KIT_SLICE_COUNT_INC).performClick() }
    fun decrementSliceCount() = apply { tag(TestTags.KIT_SLICE_COUNT_DEC).performClick() }
    fun tapSlicePad(rank: Int) = apply { tag(TestTags.kitSlicePad(rank)).performClick() }
    fun assertSliceCount(count: Int) = apply {
        // A dedicated tag avoids ambiguity with a filled pad's rank-overlay digit, which reads
        // identically to the padded readout for any double-digit count (10, 11, 12).
        tag(TestTags.KIT_SLICE_COUNT_READOUT).assert(hasText(count.toString().padStart(2, '0')))
    }

    // Pick / staged loop panel — inside the chop content's scrollable column, so bring it into
    // view first; on a short viewport it can sit below the fold beneath the slice-pad selector.
    // The panel's Column is itself clickable, so its label Text merges into its own semantics
    // node (same reasoning as the push button) — assert directly on the tagged node.
    fun assertPickPanelVisible() = apply {
        tag(TestTags.KIT_PICK_PANEL).performScrollTo().assertIsDisplayed()
    }
    fun assertPickPrompt(text: String) = apply {
        tag(TestTags.KIT_PICK_PANEL).performScrollTo().assert(hasText(text, substring = true))
    }

    // Push button — label is dynamic ("PUSH TO DEVICE · N SLICES → A", "PICK LOOP", etc.). The
    // button's clickable Box merges its label Text into its own semantics node, so assert the
    // substring directly on the tagged node rather than searching for a separate descendant.
    fun clickPush() = apply { tag(TestTags.KIT_PUSH_BUTTON).performClick() }
    fun assertPushLabel(substring: String) = apply {
        tag(TestTags.KIT_PUSH_BUTTON).assert(hasText(substring, substring = true))
    }

    // Chop progress — status text carries dynamic counts ("UPLOADING · 02 / 04",
    // "01 FAILED · 0 OK", "DONE · 04 / 04"), so assert/wait on a substring bucket.
    fun assertProgressPadVisible(rank: Int) = apply {
        tag(TestTags.kitProgressPad(rank)).assertIsDisplayed()
    }
    fun waitForProgressStatus(bucket: String, timeoutMillis: Long = 10_000) = apply {
        rule.waitUntil(timeoutMillis) {
            try {
                rule.onNode(
                    hasText(bucket, substring = true) and hasAnyAncestor(hasTestTag(TestTags.KIT_PROGRESS_STATUS)),
                ).assertIsDisplayed()
                true
            } catch (e: AssertionError) {
                false
            }
        }
    }

    // KitBuilder smoke (KIT mode content) — deliberately minimal, out of scope beyond this.
    fun assertKitBuilderContentVisible() = apply {
        text("KIT CANVAS · TAP A PAD").assertIsDisplayed()
    }
}

File: iOSApp/EP133SampleToolUITests/Robots/KitRobot.swift
import XCTest

/// Robot for the SAMPLES (Kit) screen — the loop chopper's chrome plus a smoke check that KIT
/// mode renders the embedded Kit Builder. XCUITest port of `KitRobot.kt`.
///
/// The Kit Builder's own internals (pack browser, pad canvas) live in `KitBuilderRobot`.
final class KitRobot: BaseRobot {

    /// The chop-mode slice counter is the screen's stable unique anchor.
    @discardableResult
    func assertModeChipsVisible() -> KitRobot {
        assertTextDisplayed("SLICE COUNT")
        return self
    }

    // ── Mode toggle — EP133SectionLabel uppercases, so match the rendered case ──
    @discardableResult
    func selectChopMode() -> KitRobot {
        hittableElement(TestTags.KIT_MODE_CHOP).tap()
        return self
    }

    @discardableResult
    func selectKitMode() -> KitRobot {
        hittableElement(TestTags.KIT_MODE_KIT).tap()
        return self
    }

    @discardableResult
    func assertChopModeVisible() -> KitRobot {
        assertTextDisplayed("LOOP CHOPPER")
        return self
    }

    @discardableResult
    func assertKitModeVisible() -> KitRobot {
        assertTextDisplayed("KIT BUILDER")
        return self
    }

    // ── Group + choke bar (shared with Pads via TestTags.groupChip) ──
    @discardableResult
    func selectGroup(_ group: String) -> KitRobot {
        hittableElement(TestTags.groupChip(group)).tap()
        return self
    }

    @discardableResult
    func toggleChoke() -> KitRobot {
        hittableElement(TestTags.GROUP_CHOKE_TOGGLE).tap()
        return self
    }

    /// The choke toggle renders "ON"/"OFF" as its own text; read it off the tagged element's label
    /// to disambiguate from any other ON/OFF text that may be mounted on a hidden tab.
    @discardableResult
    func assertChokeOn() -> KitRobot {
        assertChokeState("ON")
        return self
    }

    @discardableResult
    func assertChokeOff() -> KitRobot {
        assertChokeState("OFF")
        return self
    }

    /// The choke block surfaces three static texts under the same tag ("CHOKE GROUP", a hint, and
    /// the ON/OFF state). Assert the ON/OFF readout specifically by matching that label.
    private func assertChokeState(_ state: String) {
        let readout = app.staticTexts
            .matching(identifier: TestTags.GROUP_CHOKE_TOGGLE)
            .matching(NSPredicate(format: "label ==[c] %@", state))
            .firstMatch
        XCTAssertTrue(
            readout.waitForExistence(timeout: BaseRobot.defaultTimeout),
            "choke toggle should read '\(state)'")
    }

    // ── Slice count selector ──
    @discardableResult
    func incrementSliceCount() -> KitRobot {
        hittableElement(TestTags.KIT_SLICE_COUNT_INC).tap()
        return self
    }

    @discardableResult
    func decrementSliceCount() -> KitRobot {
        hittableElement(TestTags.KIT_SLICE_COUNT_DEC).tap()
        return self
    }

    @discardableResult
    func tapSlicePad(_ rank: Int) -> KitRobot {
        hittableElement(TestTags.kitSlicePad(rank)).tap()
        return self
    }

    /// A dedicated readout tag avoids ambiguity with a filled pad's rank-overlay digit; the readout
    /// text is the zero-padded count ("08").
    @discardableResult
    func assertSliceCount(_ count: Int) -> KitRobot {
        let readout = waitForTag(TestTags.KIT_SLICE_COUNT_READOUT)
        let expected = String(format: "%02d", count)
        XCTAssertEqual(readout.label, expected, "slice count should read '\(expected)'")
        return self
    }

    // ── Pick / staged loop panel ──
    @discardableResult
    func assertPickPanelVisible() -> KitRobot {
        assertTagDisplayed(TestTags.KIT_PICK_PANEL)
        return self
    }

    /// The panel's hint/headline render as their own text nodes on iOS (SwiftUI doesn't merge the
    /// container's children the way Compose's clickable Column does), so assert the text directly.
    @discardableResult
    func assertPickPrompt(_ text: String) -> KitRobot {
        // Match a static text that contains the prompt substring (headline is an exact string;
        // hints may be longer).
        let predicate = NSPredicate(format: "label CONTAINS[c] %@", text)
        let match = app.staticTexts.containing(predicate).firstMatch
        XCTAssertTrue(
            match.waitForExistence(timeout: BaseRobot.defaultTimeout),
            "expected pick panel to show '\(text)'")
        return self
    }

    // ── Push button ──
    @discardableResult
    func clickPush() -> KitRobot {
        hittableElement(TestTags.KIT_PUSH_BUTTON).tap()
        return self
    }

    /// The push button's label is uppercased and dynamic ("PUSH TO DEVICE · N SLICES → A",
    /// "PICK PACK FOLDER", …); assert a substring on the tagged button's accessibility label.
    @discardableResult
    func assertPushLabel(_ substring: String) -> KitRobot {
        let button = waitForTag(TestTags.KIT_PUSH_BUTTON)
        XCTAssertTrue(
            button.label.localizedCaseInsensitiveContains(substring),
            "push button should contain '\(substring)' (was '\(button.label)')")
        return self
    }

    // ── Chop progress ──
    @discardableResult
    func assertProgressPadVisible(_ rank: Int) -> KitRobot {
        assertTagDisplayed(TestTags.kitProgressPad(rank))
        return self
    }

    @discardableResult
    func waitForProgressStatus(_ bucket: String, timeout: TimeInterval = BaseRobot.defaultTimeout) -> KitRobot {
        let status = element(TestTags.KIT_PROGRESS_STATUS)
        XCTAssertTrue(status.waitForExistence(timeout: timeout), "expected the chop progress status")
        // The status text carries dynamic counts; the tagged node's label buckets the state.
        let predicate = NSPredicate(format: "label CONTAINS[c] %@", bucket)
        let matched = XCTNSPredicateExpectation(predicate: predicate, object: status)
        XCTAssertEqual(
            XCTWaiter().wait(for: [matched], timeout: timeout), .completed,
            "chop progress status should reach '\(bucket)'")
        return self
    }

    // ── Kit Builder smoke (KIT mode content) ──
    @discardableResult
    func assertKitBuilderContentVisible() -> KitRobot {
        assertTextDisplayed("KIT CANVAS · TAP A PAD")
        return self
    }
}
